File: tests/Web.Tests.Playwright/tests/pages/base.page.ts
import { Page, Locator } from '@playwright/test';

/**
 * Base Page Object Model
 * Contains common methods and locators shared across all pages
 */
export class BasePage {
  readonly page: Page;
  readonly navigation: Locator;
  readonly footer: Locator;
  readonly homeLink: Locator;
  readonly aboutLink: Locator;
  readonly contactLink: Locator;
  readonly articlesLink: Locator;
  readonly categoriesLink: Locator;
  readonly profileLink: Locator;
  readonly adminLink: Locator;

  constructor(page: Page) {
    this.page = page;
    this.navigation = page.locator('nav');
    this.footer = page.locator('footer');
    
    // Navigation links
    this.homeLink = page.locator('a[href="/"]');
    this.aboutLink = page.locator('a[href="/about"]');
    this.contactLink = page.locator('a[href="/contact"]');
    this.articlesLink = page.locator('a[href*="/articles"]');
    this.categoriesLink = page.locator('a[href*="/categories"]');
    this.profileLink = page.locator('a[href="/profile"]');
    this.adminLink = page.locator('a[href="/admin"]');
  }

  /**
   * Navigate to a specific URL
   */
  async goto(path: string = '/'): Promise<void> {
    await this.page.goto(path);
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Get the page title
   */
  async getTitle(): Promise<string> {
    return await this.page.title();
  }

  /**
   * Check if the navigation is visible
   */
  async isNavigationVisible(): Promise<boolean> {
    return await this.navigation.isVisible();
  }

  /**
   * Check if the footer is visible
   */
  async isFooterVisible(): Promise<boolean> {
    return await this.footer.isVisible();
  }

  /**
   * Click on Home link
   */
  async clickHome(): Promise<void> {
    await this.homeLink.first().click();
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Click on About link
   */
  async clickAbout(): Promise<void> {
    await this.aboutLink.first().click();
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Click on Contact link
   */
  async clickContact(): Promise<void> {
    await this.contactLink.first().click();
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Get the current URL
   */
  getCurrentUrl(): string {
    return this.page.url();
  }

  /**
   * Wait for specific text to appear
   */
  async waitForText(text: string): Promise<void> {
    await this.page.waitForSelector(`text=${text}`, { timeout: 10000 });
  }

  /**
   * Check if error message is displayed
   */
  async hasErrorMessage(): Promise<boolean> {
    const errorLocator = this.page.locator('[role="alert"], .alert-danger, .error-message');
    return await errorLocator.count() > 0;
  }

  /**
   * Get error message text
   */
  async getErrorMessage(): Promise<string> {
    const errorLocator = this.page.locator('[role="alert"], .alert-danger, .error-message');
    return await errorLocator.first().textContent() || '';
  }
}

File: tests/Web.Tests.Playwright/PageObjects/BasePage.cs
namespace Web.Tests.Playwright.PageObjects;

/// <summary>
/// Base Page Object Model
/// Contains common methods and locators shared across all pages
/// </summary>
public class BasePage
{
    protected readonly IPage Page;
    protected readonly ILocator Navigation;
    protected readonly ILocator Footer;
    protected readonly ILocator HomeLink;
    protected readonly ILocator AboutLink;
    protected readonly ILocator ContactLink;
    protected readonly ILocator ArticlesLink;
    protected readonly ILocator CategoriesLink;
    protected readonly ILocator ProfileLink;
    protected readonly ILocator AdminLink;

    public BasePage(IPage page)
    {
        Page = page;
        Navigation = page.Locator("nav");
        Footer = page.Locator("footer");
        
        // Navigation links
        HomeLink = page.Locator("a[href='/']");
        AboutLink = page.Locator("a[href='/about']");
        ContactLink = page.Locator("a[href='/contact']");
        ArticlesLink = page.Locator("a[href*='/articles']");
        CategoriesLink = page.Locator("a[href*='/categories']");
        ProfileLink = page.Locator("a[href='/profile']");
        AdminLink = page.Locator("a[href='/admin']");
    }

    /// <summary>
    /// Navigate to a specific URL
    /// </summary>
    public virtual async Task GotoAsync(string path = "/")
    {
        await Page.GotoAsync(path);
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Get the page title
    /// </summary>
    public async Task<string> GetTitleAsync()
    {
        return await Page.TitleAsync();
    }

    /// <summary>
    /// Check if the navigation is visible
    /// </summary>
    public async Task<bool> IsNavigationVisibleAsync()
    {
        return await Navigation.IsVisibleAsync();
    }

    /// <summary>
    /// Check if the footer is visible
    /// </summary>
    public async Task<bool> IsFooterVisibleAsync()
    {
        return await Footer.IsVisibleAsync();
    }

    /// <summary>
    /// Click on Home link
    /// </summary>
    public async Task ClickHomeAsync()
    {
        await HomeLink.First.ClickAsync();
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Click on About link
    /// </summary>
    public async Task ClickAboutAsync()
    {
        await AboutLink.First.ClickAsync();
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Click on Contact link
    /// </summary>
    public async Task ClickContactAsync()
    {
        await ContactLink.First.ClickAsync();
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Get the current URL
    /// </summary>
    public string GetCurrentUrl()
    {
        return Page.Url;
    }

    /// <summary>
    /// Wait for specific text to appear
    /// </summary>
    public async Task WaitForTextAsync(string text)
    {
        await Page.WaitForSelectorAsync($"text={text}", new() { Timeout = 10000 });
    }

    /// <summary>
    /// Check if error message is displayed
    /// </summary>
    public async Task<bool> HasErrorMessageAsync()
    {
        var errorLocator = Page.Locator("[role='alert'], .alert-danger, .error-message");
        return await errorLocator.CountAsync() > 0;
    }

    /// <summary>
    /// Get error message text
    /// </summary>
    public async Task<string> GetErrorMessageAsync()
    {
        var errorLocator = Page.Locator("[role='alert'], .alert-danger, .error-message");
        return await errorLocator.First.TextContentAsync() ?? string.Empty;
    }
}
